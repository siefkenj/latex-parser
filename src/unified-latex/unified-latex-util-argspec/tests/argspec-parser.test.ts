import { VFile } from "unified-lint-rule/lib";
import util from "util";
import * as argspecParser from "..";

/* eslint-env jest */

// Make console.log pretty-print by default
const origLog = console.log;
console.log = (...args) => {
    origLog(...args.map((x) => util.inspect(x, false, 10, true)));
};

function removeWhitespace(x: string) {
    return x.replace(/\s+/g, "");
}

describe("unified-latex-util-argspec", () => {
    let value: string | undefined;
    let file: VFile | undefined;

    const SPEC_STRINGS = [
        "",
        "o m",
        "o m o !o m",
        "!o r() m",
        "O{somedefault} m o",
        "m e{^}",
        "m e{_^}",
        "s m",
        "v!",
        "d++ D--{def}",
        "O{nested{defaults}}",
        "m ta o o",
    ];

    for (const spec of SPEC_STRINGS) {
        it(`parses xparse argument specification string "${spec}"`, () => {
            const ast = argspecParser.parse(spec);
            expect(ast).toMatchSnapshot();
            expect(argspecParser.printRaw(ast, true)).toEqual(spec);
        });
    }
});
