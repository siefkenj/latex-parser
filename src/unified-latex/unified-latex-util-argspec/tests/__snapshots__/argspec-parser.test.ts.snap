// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unified-latex-util-argspec parses xparse argument specification string "!o r() m" 1`] = `
Array [
  Object {
    "closeBrace": "]",
    "noLeadingWhitespace": true,
    "openBrace": "[",
    "type": "optional",
  },
  Object {
    "closeBrace": ")",
    "openBrace": "(",
    "type": "mandatory",
  },
  Object {
    "closeBrace": "}",
    "openBrace": "{",
    "type": "mandatory",
  },
]
`;

exports[`unified-latex-util-argspec parses xparse argument specification string "" 1`] = `Array []`;

exports[`unified-latex-util-argspec parses xparse argument specification string "O{nested{defaults}}" 1`] = `
Array [
  Object {
    "closeBrace": "]",
    "defaultArg": Object {
      "content": Array [
        "n",
        "e",
        "s",
        "t",
        "e",
        "d",
        Object {
          "content": Array [
            "d",
            "e",
            "f",
            "a",
            "u",
            "l",
            "t",
            "s",
          ],
          "type": "group",
        },
      ],
      "type": "group",
    },
    "openBrace": "[",
    "type": "optional",
  },
]
`;

exports[`unified-latex-util-argspec parses xparse argument specification string "O{somedefault} m o" 1`] = `
Array [
  Object {
    "closeBrace": "]",
    "defaultArg": Object {
      "content": Array [
        "s",
        "o",
        "m",
        "e",
        "d",
        "e",
        "f",
        "a",
        "u",
        "l",
        "t",
      ],
      "type": "group",
    },
    "openBrace": "[",
    "type": "optional",
  },
  Object {
    "closeBrace": "}",
    "openBrace": "{",
    "type": "mandatory",
  },
  Object {
    "closeBrace": "]",
    "openBrace": "[",
    "type": "optional",
  },
]
`;

exports[`unified-latex-util-argspec parses xparse argument specification string "d++ D--{def}" 1`] = `
Array [
  Object {
    "closeBrace": "+",
    "openBrace": "+",
    "type": "optional",
  },
  Object {
    "closeBrace": "-",
    "defaultArg": Object {
      "content": Array [
        "d",
        "e",
        "f",
      ],
      "type": "group",
    },
    "openBrace": "-",
    "type": "optional",
  },
]
`;

exports[`unified-latex-util-argspec parses xparse argument specification string "m e{^}" 1`] = `
Array [
  Object {
    "closeBrace": "}",
    "openBrace": "{",
    "type": "mandatory",
  },
  Object {
    "embellishmentTokens": Array [
      "^",
    ],
    "type": "embellishment",
  },
]
`;

exports[`unified-latex-util-argspec parses xparse argument specification string "m e{_^}" 1`] = `
Array [
  Object {
    "closeBrace": "}",
    "openBrace": "{",
    "type": "mandatory",
  },
  Object {
    "embellishmentTokens": Array [
      "_",
      "^",
    ],
    "type": "embellishment",
  },
]
`;

exports[`unified-latex-util-argspec parses xparse argument specification string "m ta o o" 1`] = `
Array [
  Object {
    "closeBrace": "}",
    "openBrace": "{",
    "type": "mandatory",
  },
  Object {
    "token": "a",
    "type": "optionalToken",
  },
  Object {
    "closeBrace": "]",
    "openBrace": "[",
    "type": "optional",
  },
  Object {
    "closeBrace": "]",
    "openBrace": "[",
    "type": "optional",
  },
]
`;

exports[`unified-latex-util-argspec parses xparse argument specification string "o m o !o m" 1`] = `
Array [
  Object {
    "closeBrace": "]",
    "openBrace": "[",
    "type": "optional",
  },
  Object {
    "closeBrace": "}",
    "openBrace": "{",
    "type": "mandatory",
  },
  Object {
    "closeBrace": "]",
    "openBrace": "[",
    "type": "optional",
  },
  Object {
    "closeBrace": "]",
    "noLeadingWhitespace": true,
    "openBrace": "[",
    "type": "optional",
  },
  Object {
    "closeBrace": "}",
    "openBrace": "{",
    "type": "mandatory",
  },
]
`;

exports[`unified-latex-util-argspec parses xparse argument specification string "o m" 1`] = `
Array [
  Object {
    "closeBrace": "]",
    "openBrace": "[",
    "type": "optional",
  },
  Object {
    "closeBrace": "}",
    "openBrace": "{",
    "type": "mandatory",
  },
]
`;

exports[`unified-latex-util-argspec parses xparse argument specification string "s m" 1`] = `
Array [
  Object {
    "type": "optionalStar",
  },
  Object {
    "closeBrace": "}",
    "openBrace": "{",
    "type": "mandatory",
  },
]
`;

exports[`unified-latex-util-argspec parses xparse argument specification string "v!" 1`] = `
Array [
  Object {
    "closeBrace": "!",
    "openBrace": "!",
    "type": "verbatim",
  },
]
`;
